index-handlers/src/main/java/no/sikt/nva/nvi/index/utils/InstitutionReportGenerator.java

```diff
@@ -0,0 +1,129 @@
+    private final String year;
+    private final URI topLevelOrganization;
+
+    public InstitutionReportGenerator(SearchClient<NviCandidateIndexDocument> searchClient,
+                                      String year,
+                                      URI topLevelOrganization) {
+        this.searchClient = searchClient;
+        this.year = year;
+        this.topLevelOrganization = topLevelOrganization;
+    }
+
+    public ExcelWorkbookGenerator generateReport() {
+        var nviCandidates = fetchNviCandidates();
+        var data = nviCandidates.stream().flatMap(this::generateDataRows).toList();
+        return new ExcelWorkbookGenerator(INSTITUTION_REPORT_HEADERS, data);
+    }
+
+    private Stream<List<String>> generateDataRows(NviCandidateIndexDocument candidate) {
+        return generateReportRowsForCandidate(candidate).stream();
+    }
+
+    private List<NviCandidateIndexDocument> fetchNviCandidates() {
+        return attempt(() -> searchClient.search(buildSearchRequest())).orElseThrow()
+                   .hits()
+                   .hits()
+                   .stream()
+                   .map(Hit::source)
+                   .toList();
+    }
+
+    private CandidateSearchParameters buildSearchRequest() {
+        return CandidateSearchParameters.builder()
+                   .withYear(year)
+                   .withTopLevelCristinOrg(topLevelOrganization)
+                   .build();
+    }
+
+    private List<List<String>> generateReportRowsForCandidate(NviCandidateIndexDocument candidate) {
+        return attempt(() -> generateReportRowsForContributorAffiliations(candidate)).map(Stream::toList)
+                   .orElseThrow(failure -> {
+                       logFailure(failure.getException(), candidate);
+                       return (RuntimeException) failure.getException();
+                   });
+    }
+
+    private Stream<List<String>> generateReportRowsForContributorAffiliations(NviCandidateIndexDocument candidate) {
+        return candidate.getNviContributors().stream()
+                   .flatMap(nviContributor -> generateRowsForContributorAffiliations(candidate, nviContributor));
+    }
+
+    private void logFailure(Exception exception, NviCandidateIndexDocument candidate) {
+        logger.error("Failed to generate report lines for candidate: {}. Error {}", candidate.id(),
+                     getStackTrace(exception));
+    }
+
+    private Stream<List<String>> generateRowsForContributorAffiliations(NviCandidateIndexDocument candidate,
+                                                                        NviContributor nviContributor) {
+        return nviContributor.getAffiliationsPartOfOrEqualTo(topLevelOrganization)
+                   .map(affiliation -> generateRow(candidate, nviContributor, affiliation));
+    }
+
+    private List<String> generateRow(NviCandidateIndexDocument candidate, NviContributor nviContributor,
```

Comment:
Would it be worth it to extract this to a record or something (with a toList() if needed)? This relies on insertion order in a list of arbitrary strings, so it could be hard to verify that the structure is correct (since it should match the headers).

Reply:
See refactor, moved report representation creation to `NviCandidateIndexDocument`, returning a `List` of key-value-map, which can be ordered by order of enum `InstitutionReportHeader`